### PR TITLE
Add high-level control script for post-processing

### PR DESCRIPTION
**Pull request by pineapple-cat**
Addresses #104.

- [x] Small update to display name parsing (unrelated QOL update)
- [x] Command line argument parsing
- [x] Get perflog information to pass to plotting scripts
- [x] Display the relevant dataframe columns used
- [x] Unit tests for dataframe collation

### CODE REVIEW

File: post-processing/post_processing.py [hunk 1]
```diff
@@ -1,18 +1,89 @@
...
+        # identify a relevant subset of the dataframe to pass on
+        # TODO
+        columns = ["test_name", "system", "partition", "environ"]
+        for fom in figures_of_merit:
+            value = fom + "_value"
+            unit = fom + "_unit"
+            if (value in df.columns) & (unit in df.columns):
+                columns.extend([value, unit])
+            else:
+                if self.debug:
```
Comment (pineapple-cat) [inline post-processing/post_processing.py:59]: Should be visible even without debug flag (print all available columns and throw an error).

File: post-processing/post_processing.py [hunk 2]
```diff
@@ -1,18 +1,89 @@
...
+            print("Found log files:")
+            for log in log_files:
+                print("-", log)
+
+        df = pd.DataFrame()
+        # put all perflog information in one dataframe
+        for file in log_files:
+            try:
+                temp = read_perflog(file)
+                df = pd.concat([df, temp], ignore_index=True)
```
Comment (pineapple-cat) [inline post-processing/post_processing.py:42]: Double-check how this works with introducing new columns.

File: post-processing/post_processing.py [hunk 3]
```diff
@@ -1,18 +1,89 @@
...
+            except KeyError as e:
+                if self.debug:
+                    print("Discarding %s:" %os.path.basename(file), type(e).__name__ + ":", e.args[0], e.args[1])
+
+        if df.empty:
+            raise FileNotFoundError(errno.ENOENT, "Could not find a valid perflog in path", path)
+
```
Comment (pineapple-cat) [inline post-processing/post_processing.py:49]: Add a loop for optional filtering of rows (e.g. by date, by system, etc.). Need to have this as an input to the run function (could mean some columns that were filtered by are no longer necessary).

If this is empty (e.g. no rows passed the filters), throw an error.

**Comment by pineapple-cat** [comment]
Implemented suggestions:

- [ ] Change `run_post_processing` input to get a path to a config file containing:
   a.) user-selected columns
   b.) filter conditions for row selection
- [ ] Make invalid columns throw an error + check there are at least 3 columns specified:
   a.) figure of merit value
   b.) figure of merit unit
   c.) something to plot the figure of merit against
- [ ] Add optional row filtering function + throw an error if this results in an empty DataFrame
- [ ] Add test for concatenating DataFrames with different columns

File: post-processing/post_processing.py [hunk 4]
```diff
@@ -43,33 +49,107 @@ def run_post_processing(self, path, figures_of_merit):
...
+        # evaluate expression and extract dataframe mask
+        if value is None:
+            if operator == op.eq:
+                mask = df[column].isnull()
+            else:
+                mask = df[column].notnull()
+        else:
+            try:
+                # dataframe column is interpreted as the same type as the supplied value
+                mask = operator(df[column].astype(type(value)), value)
```
Comment (ilectra) [inline post-processing/post_processing.py:132]: There's a slight issue here, if the type of values of the column doesn't agree with the type of the filter value from the YAML (eg. int vs float). But this is part of a larger issue, that will be resolved separately, see #176

File: post-processing/post_processing_config.yaml [hunk 5]
```diff
@@ -0,0 +1,23 @@
+# Plot title
```
[the remark this answers is not in the document]
Comment (pineapple-cat) [reply post-processing/post_processing_config.yaml:1]: Opened #178 for this

File: post-processing/post_processing.py [hunk 6]
```diff
@@ -1,41 +1,179 @@
...
+
+        invalid_columns = []
+        # check for invalid columns
+        for col in columns:
+            if col not in df.columns:
+                invalid_columns.append(col)
+        if invalid_columns:
+            raise KeyError("Could not find columns", invalid_columns)
+
+        filters = config["filters"]
```
Comment (ilectra) [inline post-processing/post_processing.py:70]: A check just like the one you have above, to make sure the needed columns for the filters are included in the dataframe, has to be added here. Same for the datasets.
Note that this check has to be here and not in the `read_config`, as is compares to the df.

File: post-processing/post_processing.py [hunk 7]
```diff
@@ -1,41 +1,179 @@
...
+        mask = pd.Series(df.index.notnull())
+        # filter rows
+        if filters:
+            mask = reduce(op.and_, (self.row_filter(f, df) for f in filters))
+        if df[mask].empty:
+            raise pd.errors.EmptyDataError("Filtered dataframe is empty", df[mask].index)
+
```
Comment (ilectra) [inline post-processing/post_processing.py:77]: If datasets have been given in the config
- their columns should be added to the `columns` here
- the number of rows after the filtering should be = (number of x points)x(number of datasets). If no datasets are given, then (number of rows)=(number of x points).

File: post-processing/post_processing_config.yaml [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+
+# Optional setting to display several plots in the same graph
+# Dataset format: [column_name, value]
+datasets: [["system", "default"]]
+
+# Axes must have a value specified with a column name
+# Units can either be specified with a column name
+# or a custom label (including null (i.e. None) if units are N/A)
+
+x_axis:
```
Comment (ilectra) [inline post-processing/post_processing_config.yaml:23]: More checks to be added to `read_config`: the value and units for both axis should be the same as the `columns` given above.

File: post-processing/post_processing.py (hunk 7 above)
Comment (ilectra) [reply post-processing/post_processing.py:77]: I should say "number of _unique_ x points". Maybe read that column into a set to get this number?

File: post-processing/post_processing_config.yaml (hunk 8 above)
Comment (pineapple-cat) [reply post-processing/post_processing_config.yaml:23]: Depending on what we need additional columns for, we could potentially remove the columns from config entirely and just internally compile a necessary columns list based on axis values, units, and datasets.

**Review by ilectra — APPROVED** [review]

**Pull request merged**